pkg/porter/porter.go

```diff
@@ -47,22 +52,47 @@ type Porter struct {
 	Plugins       plugins.PluginProvider
 	CNAB          cnabprovider.CNABProvider
 	Secrets       secrets.Store
-	Storage       storage.Provider
 	Signer        signing.Signer
+
+	onClose []func() error
+
+	// Deprecated: Use the individual storage providers in the Porter struct instead
+	// This is only here for backwards compatibility where MongoDB was the only storage provider.
+	Storage storage.Provider
+}
+
+// Options for configuring a new Porter client passed to NewWith.
+type Options struct {
+	Config        *config.Config                // Optional. Defaults to a config.New.
+	Secrets       secrets.Store                 // Optional. Defaults to a secrets.NewPluginAdapter(secretsplugin.NewStore).
+	Signer        signing.Signer                // Optional. Defaults to a signing.NewPluginAdapter(signingplugin.NewSigner).
+	Installations storage.InstallationProvider  // Optional. Defaults to a storage.NewInstallationStore for MongoDB or storage.NewInstallationStoreSQL for SQL.
+	Credentials   storage.CredentialSetProvider // Optional. Defaults to a storage.NewCredentialStore for MongoDB or storage.NewCredentialStoreSQL for SQL.
+	Parameters    storage.ParameterSetProvider  // Optional. Defaults to a storage.NewParameterStore for MongoDB or storage.NewParameterStoreSQL for SQL.
+	Sanitizer     *storage.Sanitizer            // Optional. Defaults to a storage.NewSanitizer for MongoDB or storage.NewSanitizer for SQL.
+}
+
+// NewWith creates a new Porter client with useful defaults that can be overridden with the provided options.
+func NewWith(opts Options) *Porter {
+	return newWith(opts.Config, opts.Installations, opts.Credentials, opts.Parameters, opts.Secrets, opts.Signer, opts.Sanitizer, nil)
 }
 
 // New porter client, initialized with useful defaults.
+// Use NewWith to override the defaults.
 func New() *Porter {
-	c := config.New()
-	storage := storage.NewPluginAdapter(storageplugin.NewStore(c))
-	secretStorage := secrets.NewPluginAdapter(secretsplugin.NewStore(c))
-	signer := signing.NewPluginAdapter(signingplugin.NewSigner(c))
-	return NewFor(c, storage, secretStorage, signer)
+	// differ full initialization to Connect
+	return newWith(nil, nil, nil, nil, nil, nil, nil, nil)
 }
 
-func NewFor(c *config.Config, store storage.Store, secretStorage secrets.Store, signer signing.Signer) *Porter {
-	cache := cache.New(c)
-
+// NewFor creates a new Porter client with the provided configuration and storage backend.
+//
+// Deprecated: Use NewWith instead. NewFor does not support newer sql storage backends.
```

Comment:
I mean, I'm fine to remove it since we already say [Porter does not guarantee backwards compatibility in its library](https://porter.sh/docs/references/library/). :)